### ISSUE TITLE
lerna pack <...args> to generate local tarballs

### ISSUE THREAD

**Issue by indexzero** [lerna/lerna#2134]
> Did a fair amount of searching within open & closed issues before opening this. If I have over looked something obvious please let me know.

Hi. Long time lurker, first time poster. Very comfortable with `npm pack` behavior:

```
$ npm pack
npm info it worked if it ends with ok
npm info using npm@6.4.1
npm info using node@v10.15.3
npm info lifecycle triple-beam@1.3.0~prepublish: triple-beam@1.3.0
npm info lifecycle triple-beam@1.3.0~prepare: triple-beam@1.3.0
npm info lifecycle triple-beam@1.3.0~prepack: triple-beam@1.3.0
npm info lifecycle triple-beam@1.3.0~postpack: triple-beam@1.3.0
npm notice 
npm notice 📦  triple-beam@1.3.0
npm notice === Tarball Contents === 
npm notice 878B  package.json    
npm notice 136B  .eslintrc       
npm notice 25B   .gitattributes  
npm notice 220B  .travis.yml     
npm notice 374B  CHANGELOG.md    
npm notice 1.1kB index.js        
npm notice 1.1kB LICENSE         
npm notice 689B  README.md       
npm notice 2.5kB test.js         
npm notice 616B  config/cli.js   
npm notice 538B  config/index.js 
npm notice 516B  config/npm.js   
npm notice 556B  config/syslog.js
npm notice === Tarball Details === 
npm notice name:          triple-beam                             
npm notice version:       1.3.0                                   
npm notice filename:      triple-beam-1.3.0.tgz                   
npm notice package size:  3.4 kB                                  
npm notice unpacked size: 9.2 kB                                  
npm notice shasum:        a595214c7298db8339eeeee083e4d10bd8cb8dd9
npm notice integrity:     sha512-XrHUvV5HpdLmI[...]qfAbEe/QMp2Lw==
npm notice total files:   13                                      
npm notice 
triple-beam-1.3.0.tgz
npm timing npm Completed in 343ms
npm info ok 
```

Looking for similar behavior in `lerna`. Don't want to publish anything or increment any versions, just want to generate `*.tgz` files based on the current configs and CLI args passed in.

**Comment by evocateur** [comment]
I agree, this would be very helpful. Unfortunately, the implementation of how this is done (running lifecycles, replacing version strings, etc) is very tightly-coupled to the `lerna publish` command. It would be a dramatic (but welcome!) refactoring to extract into a sharable module that could be reused by a new `lerna pack` command.

**Comment by LeeeeeeM** [comment]
mark. useful instruction

**Comment by gstvribs** [comment]
A simple workaround that works for me, is by simple run npm pack inside of each package, like:

`
// packageA/package.json

"pack": "npm pack"

// monorepo/package.json

"lerna:pack": "lerna run pack",
"lerna:pack:changed": "lerna run pack --since"
`

**Comment by szul** [comment]
@gstvribs But does that resolve `lerna link convert` packages and add them back into the packed tarball?

**Comment by gstvribs** [comment]
No, you have a reason this behavior will not work, actually, I did not have any symlink package

**Comment by jabranr** [comment]
Good shout! This definitely eases packages testing.

**Comment by qfox** [comment]
Unfortunately, we wrote a lot of scripts around lerna commands but ideas can be useful if we decide to implement them to lerna.

We using something like `lerna changed` to calculate changes, then store it in artefact, then in other task we reuse it to run something like `lerna vesion` and bootstrap only affected packaged to run pack in each, then we store ready to publish packages in artefact, and then, in the last task, reuse ready packages to publish to Verdaccio (ye, we use it too).
The last one needed to precheck packages building and separate publish process from pushing commits to VCS.

Another idea was to publish packages in draft mode (or start transaction) in Verdaccio and then (if VCS accepted our version changes) just commit them, or revert if VCS failed.

Prob these would be helpful. I hope so ;-)

**Comment by ghdoergeloh** [comment]
This would be great for CI if you want to build the packages, run some other tests and only if they succeed deploy the packages.

I currently have an aws pipeline that first runs the main job to install dependencies, build TS, test, lint and so on.
After this job in another job (other docker image) sonarscanner will check for bugs using the coverage from the container before.
The last job should only publish the packages but to copy the whole folder with all installed dependencies from the first job as an artifact or to install and build the package again, does not seem so performant. Just copying all the ready build packages and only deploying those would be nice.

We do the versioning "manualy" with `lerna version` and let the pipeline just publish on a git tag (`lerna publish from-git`). So I would need something like `lerna pack from-git`.

Is there a workaround for this? Packing all and select the changed would be okay, but there is no simple way to check, which packaged changed like `from-git` does, as far as I see, but I might be wrong.

**Comment by ghdoergeloh** [comment]
Another related (old) feature request #2527

**Comment by ghdoergeloh** [comment]
So as far as I can see, the most simple solution would be, to only make this line optional

https://github.com/lerna/lerna/blob/674ffd369cf446ca2d7c2811d882a5462c8bbfd7/commands/publish/index.js#L246

(something like "--no-publish" wich would sound a bit confusing: `lerna publish --no-publish` so `lerna pack` would be more convenient)

and skip git-reset with "--no-git-reset" (which already exists).

https://github.com/lerna/lerna/blob/674ffd369cf446ca2d7c2811d882a5462c8bbfd7/commands/publish/index.js#L248-L250

afterwards we would have all changed the packages packed. With this I could run `npm publish` for all these packages in another pipeline step.

**Comment by morriswinkler** [comment]
maybe `lerna publish --pack-only` and make an alias to `lerna pack`

**Comment by T0miii** [comment]
is there an acceptable work around for the time being?

**Comment by luttje** [comment]
> is there an acceptable work around for the time being?

Depends on what you find acceptable as a workaround, and for what purpose you need to pack the repo's.

I need a workaround in order to run integration tests, so my approach is:

1. Start a local npm registry (like [verdaccio](https://verdaccio.org/)), e.g:

    ```bash
    verdaccio --listen 4000 --config ./config.yaml
    ```

    With a `config.yaml` like this:

    ```yaml
    # The storage is where the tarballs will be stored
    storage: ./storage
    uplinks:
      npmjs:
        url: https://registry.npmjs.org/
    packages:
      '@my-org/*':
        access: $all
        publish: $all
        # commented so npmjs registry is not used for our packages
        # proxy:
      '**':
        access: $all
        publish: $all
        # all other dependencies are fetched from npmjs
        proxy: npmjs
    ```

2. Have Lerna publish to that local registry using:

    ```bash
    lerna publish --canary --no-git-tag-version --no-push --force-publish --no-private --yes --registry http://localhost:4000
    ```

3. Now I can test by installing the package from that local registry:

    ```bash
    npm install @my-org/package-name --registry=http://localhost:4000
    ```

    Or you can access the published tarballs in the ./storage folder of verdaccio.

Hope this helps

**Comment by T0miii** [comment]
@luttje we usually did pack our library and before publishing or pushing we did a local install of the package in one of our other projects to test the changes, i will try this approach.

**Comment by aaronclong** [comment]
Is there any chance of getting this added? There have been a few issue created for this feature.

https://github.com/lerna/lerna/issues/2527
